### ISSUE TITLE
Estimate gas seems to fail if the transaction does a revert

### ISSUE THREAD

**Issue by axic** [MetaMask/metamask-extension#2714]
`eth_estimateGas` seems to fail if the transaction estimated will do a `revert`.

It would be nice to produce a proper message explaining it did a revert as opposed to the generic out of gas message which claims it is over 3 million gas.

**Comment by danfinlay** [comment]
This will require running a local VM, and running the estimation locally, which is something we want to do already, as per #2083.

Leaving this issue open for the specific enhancement of detecting reverts.

**Comment by axic** [comment]
Was this issue sorted? See @danfinlay's last comment.

**Comment by danfinlay** [comment]
This is a legit problem that comes from not having a local VM. Reopening unless there's a good reason to make it a new issue.

**Comment by jtakalai** [comment]
Is this related to an issue we're having, namely, instead of reporting that a transaction would revert, Metamask would say "transaction is fine, but it will take 9.7M gas"? Metamask-generated transaction here: https://etherscan.io/tx/0xa0e773128a5b2e2b2008bbfa92885c90088fbb1c55cc098ae6b68329c5c17506 replaced with one where gas limit was lower (because that one would never got into a block)

**Comment by danfinlay** [comment]
MetaMask gas estimations come from geth. Since we have no local VM, any revert-detection needs to be provided by the client via RPC before we can include it.

**Comment by Gudahtt** [comment]
I believe that Infura tells us in the `estimateGas` error response whether it was due to a revert or not. So I'm not sure revert-detection is a problem here.

We'd still need to think through how to best display this to the user though. Also the initial gas limit calculation is far removed from any UI, and this is used in a few different workflows, so this might be a little bit tangly purely for code organization reasons.

**Comment by bschorchit** [comment]
Closing in favor of: https://github.com/MetaMask/metamask-extension/issues/19692

Additionally, we have added a new alert on the redesigned confirmations (soon to be default) to callout when MM fails to estimate the gas limit.
